File: LeetcodeQuestions/n-ary_tree_preorder_traversal-589.rb
# Definition for a Node.
# class Node
#     attr_accessor :val, :children
#     def initialize(val)
#         @val = val
#         @children = []
#     end
# end

# @param {Node} root
# @return {List[int]}
def preorder(root)
    return [] if !root
    stack = [root]
    res = []
    
    while stack.length > 0
        node = stack.pop()
       
        if node.val == nil
            next
        end
        
        res.push(node.val)
        
        if node.children.length > 0
            i = node.children.length - 1
            
            while i >= 0
                child = node.children[i]
                stack.push(child)
                i -= 1
            end
        end
    end
    return res
end

File: LeetcodeQuestions/n-aryTreePreorderTraversal-589.js
/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {Node} root
 * @return {number[]}
 */
var preorder = function(root) {
    if (!root) return [];
    let stack = [root];
    let res = [];
    
    while (stack.length) {
        let node = stack.pop();
        if (node === null) continue;
        res.push(node.val);
        if (node.children.length) {
            for (let i = node.children.length - 1; i >= 0; i--) {
                stack.push(node.children[i])
            }
        }
    }
    return res;
};